### PR TITLE
Fix JSON path expressions in SQL queries.

### PR DESCRIPTION
**Pull request by amrishlal**
## Description
PR #7590 introduced a regression that caused JSON path expressions in queries to stop working. This PR fixes the regression and adds an integration test that would catch such regressions in future.
## Upgrade Notes
Does this PR prevent a zero down-time upgrade? (Assume upgrade order: Controller, Broker, Server, Minion)
* [ ] Yes (Please label as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR fix a zero-downtime upgrade introduced earlier?
* [ ] Yes (Please label this as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR otherwise need attention when creating release notes? Things to consider:
- New configuration options
- Deprecation of configurations
- Signature changes to public methods/interfaces
- New plugins added or old plugins removed
* [ ] Yes (Please label this PR as **<code>release-notes</code>** and complete the section on Release Notes)

### CODE REVIEW

**Comment by richardstartin** [comment]
This isn't a review comment but can you let me know if you intend to make any changes on the server please?

**Comment by mayankshriv** [comment]
> This isn't a review comment but can you let me know if you intend to make any changes on the server please?

+1. Also description would be nice so we can make sure any parallel work on JSON is aligned.

**Comment by kishoreg** [comment]
@amrishlal we should reconsider how we are supporting JSON syntax. a.b is a well-known syntax in SQL and it typically refers to tablename.column name. Overloading that to support a nested json path is dangerous and can result in a lot of complications down the road especially when we implement JOIN, UNNEST, etc. 

If you really need to support this within LinkedIn, I strongly suggest using some other separator. For e.g. Postgres use "->" and data bricks uses ":"

https://www.postgresql.org/docs/9.3/functions-json.html

**Comment by amrishlal** [comment]
> This isn't a review comment but can you let me know if you intend to make any changes on the server please?

@richardstartin Only BaseBrokerRequestHandler.java is being changed. Along with that I will add an integration test that will detect future breakages.

@mayankshriv @kishoreg This isn't a new feature, this is a bug fix. The JSON naming convention that is described in 

I will add a description with more detailed information. Please note that this is not a new feature and there is no change to any existing functionality. The PR is fixing a regression which crept in due to a prior commit (https://github.com/apache/pinot/pull/7590).

**Comment by amrishlal** [comment]
@kishoreg @mayankshriv @richardstartin FYI: this PR is ready for review.

**Comment by kishoreg** [comment]
can we decide on the delimiter right now? @siddharthteotia 

I really prefer not to overload . in the identifier. SQL standard is tableName.columnName. Can we something like postgres person -> age or like spark person:age

we want to support unnest, join etc and this will add a lot of complexity to maintain backward compatibility.

**Comment by amrishlal** [comment]
> can we decide on the delimiter right now? @siddharthteotia
> 
> I really prefer not to overload . in the identifier. SQL standard is tableName.columnName. Can we something like postgres person -> age or like spark person:age
> 
> we want to support unnest, join etc and this will add a lot of complexity to maintain backward compatibility.

@kishoreg We can change the delimiter if you really want, but this isn't really overloading. We are just applying the standard dot notation to go to finer level of detail, i.e going from table to column to JSON elements within the column values. Please note that dot notation will only work for columns of type JSON. Oracle JSON database already uses this notation for JSON document (so this isn't exactly without precedence). I fail to see why postgres went with "->" (at least to me it seems like an arbitary choice). A secondary issue here is that delimiters such as "->" or ":" in identifiers are not supported by calcite (but ".", "[", and "]" are), so that allows us to suffix JSON path expressions behind column name. I had discussed this issue a bit on the calcite side as well (and my understanding is that they are open to extending the notation to include json path expressions in future). In short, I think we should be ok with using "." here. Happy to discuss further if you wish.

File: pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java [hunk 1]
```diff
@@ -1614,40 +1614,48 @@ private static void fixColumnName(String rawTableName, Expression expression, Ma
    * - Case-insensitive cluster
    * - Column name in the format of [table_name].[column_name]
    */
-  private static String getActualColumnName(String rawTableName, String columnName,
+  private static String getActualColumnName(String rawTableName, String sqlColumnName,
```
Comment (Jackie-Jiang) [inline pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java:1617]: Let's not change the argument name here. If we want to preserve the name for exception message, we can either use a separate variable to store the mutated column name like the current code, or keep it in a separate local variable `String originalColumnName = columnName;`

File: pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java [hunk 2]
```diff
@@ -1614,40 +1614,48 @@ private static void fixColumnName(String rawTableName, Expression expression, Ma
...
+    // as column name.
+    String[] tableSplit = StringUtils.split(columnName, ".", 2);
+    if (tableSplit.length == 2) {
+      if ((isCaseInsensitive && rawTableName.equalsIgnoreCase(tableSplit[0])) || rawTableName.equals(tableSplit[0])) {
+        columnName = tableSplit[1];
       }
     }
+
+    // Split again to check if the column name is suffixed by a path expression.
+    String[] columnSplit = StringUtils.split(columnName, ".", 2);
```
Comment (Jackie-Jiang) [inline pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java:1634]: This won't work if the column name itself contains `.`. We should first check if the column exist, if not, try to treat it as json path expression.

File: pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/JSONClusterIntegrationTest.java [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+    record.putValue(JSON_COLUMN, jsonValue);
+    record.putValue(JSON_COLUMN_WITHOUT_INDEX, jsonValue);
+
+    return record;
+  }
+
+  @Test
+  public void testPathExpressionQueries()
+      throws Exception {
+    JsonNode queryResponse = postSqlQuery(
```
Comment (siddharthteotia) [inline pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/JSONClusterIntegrationTest.java:147]: Can you try an example where the array data itself is nested object and not a flat structure (like `\"data\": [\"k\", \"b\", \"c\", \"d\"]` ) and then use that array in filtering and selecting without specifying the index ?

**Pull request closed without merging**